timeutil: add high precision clock on Windows

tamird opened the pull request:
This gets the list of failures down to:
```
--- FAIL: TestHeartbeatHealthTransport (45.23s)
FAIL
FAIL	github.com/cockroachdb/cockroach/pkg/rpc	45.658s
--- FAIL: TestStatusLocalLogs (0.20s)
FAIL
FAIL	github.com/cockroachdb/cockroach/pkg/server	12.288s
--- FAIL: TestEval (0.05s)
FAIL
FAIL	github.com/cockroachdb/cockroach/pkg/sql/parser	1.129s
--- FAIL: TestConcurrentBatch (20.93s)
FAIL
FAIL	github.com/cockroachdb/cockroach/pkg/storage/engine	26.650s
--- FAIL: TestGetLogReader (0.03s)
    --- FAIL: TestGetLogReader/logtestexe.log (0.00s)
        --- FAIL: TestGetLogReader/logtestexe.log/restricted=true (0.00s)
        --- FAIL: TestGetLogReader/logtestexe.log/restricted=false (0.00s)
    --- FAIL: TestGetLogReader/cockroach.roach0.root.2015-09-25T19_24_19Z.00000.log (0.00s)
        --- FAIL: TestGetLogReader/cockroach.roach0.root.2015-09-25T19_24_19Z.00000.log/restricted=true (0.00s)
--- FAIL: TestGC (0.10s)
FAIL
FAIL	github.com/cockroachdb/cockroach/pkg/util/log	0.550s
--- FAIL: TestTimeConversion (0.00s)
FAIL
FAIL	github.com/cockroachdb/cockroach/pkg/util/timeutil	0.438s
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14597)
<!-- Reviewable:end -->

tamird committed:
timeutil: add high precision clock on Windows

time.Now() has a precision of about 1ms on Windows[0], which is coarse
enough to fail some of our tests. This implementation has a precision
of about 10µs, which is fine enough to trip a hastily-written test in
storage; that test is also fixed here.

[0] https://github.com/golang/go/issues/8687

Note that this implementation is significantly slower than time.Now:
```
name   old time/op  new time/op   delta
Now-2  14.0ns ± 3%  339.4ns ± 4%  +2332.32%  (p=0.000 n=9+9)
```
I suspect this is because of the explicit syscall required. time.Now
avoids the syscall by reading the system clock directly[1][2], so if
we need better performance, we can do something similar.

[1] https://github.com/golang/go/commit/74b62b48644acffd20ca5de9111048dd731c3a06
[2] https://github.com/golang/go/blob/go1.8/src/runtime/os_windows.go#L581:L624

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/util/timeutil/time_unix.go, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14597#-KgtD32xFfBgHE1imuBL:-KgtD32xFfBgHE1imuBM:b4g0htg) ([raw file](https://github.com/cockroachdb/cockroach/blob/84a03628ef01e30733830373648baabbf3205d8e/pkg/util/timeutil/time_unix.go#L23)):*
> ```Go
> )
> 
> func now() time.Time {
> ```

There is now an extra-level of indirection: `nowFunc -> now -> time.Now`. Hopefully the compiler eliminates that, but perhaps this should be:

```
var nowImpl = time.Now
```

And the windows version:

```
var nowImpl = func() time.Time {
  ...
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14597#-:-KgtDRY7LmNBw9u2R1x-:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/util/timeutil/time_unix.go, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14597#-KgtD32xFfBgHE1imuBL:-KgtDnBy0AcpbQTcA7jZ:b-xxtow4) ([raw file](https://github.com/cockroachdb/cockroach/blob/84a03628ef01e30733830373648baabbf3205d8e/pkg/util/timeutil/time_unix.go#L23)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

There is now an extra-level of indirection: `nowFunc -> now -> time.Now`. Hopefully the compiler eliminates that, but perhaps this should be:

```
var nowImpl = time.Now
```

And the windows version:

```
var nowImpl = func() time.Time {
  ...
}
```
</blockquote></details>

I think the compiler is actually less likely to eliminate that, because of the `var` on `nowImpl`; on the other hand, the current code is provably constant.

I wrote a benchmark though, and you turned out to be right (new is with your suggestion):
```
name   old time/op  new time/op  delta
Now-4  13.4ns ± 5%  11.9ns ± 1%  -10.88%  (p=0.000 n=10+9)
```

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14597)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/queue_test.go, line 775 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14597#-KgtJr2qkx2KDmTsEY4_:-KgtJr2qkx2KDmTsEY4a:bk21594) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2f90817b0569a18ced72bced36c78d45119baf4/pkg/storage/queue_test.go#L775)):*
> ```Go
> 		}
> 		if min, v := bq.queueConfig.processTimeout, bq.processingNanos.Count(); v < min.Nanoseconds() {
> 			return errors.Errorf("expected >= %s in processing time; got %s", min, time.Duration(v))
> ```

This seems unrelated to this PR. Revert?

---

*[pkg/util/timeutil/time_unix.go, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14597#-KgtD32xFfBgHE1imuBL:-KgtJXiK2b4bRyuI6-9Y:bv13rv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/84a03628ef01e30733830373648baabbf3205d8e/pkg/util/timeutil/time_unix.go#L23)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think the compiler is actually less likely to eliminate that, because of the `var` on `nowImpl`; on the other hand, the current code is provably constant.

I wrote a benchmark though, and you turned out to be right (new is with your suggestion):
```
name   old time/op  new time/op  delta
Now-4  13.4ns ± 5%  11.9ns ± 1%  -10.88%  (p=0.000 n=10+9)
```

Done.
</blockquote></details>

My suggestion eliminates `now` from the call chain: `nowFunc = nowImpl` becomes an assignment of one function pointer to another so `nowFunc()` directly invokes `time.Now()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14597)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/storage/queue_test.go, line 775 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14597#-KgtJr2qkx2KDmTsEY4_:-KgtK2PbeFfTAMazqCBH:b-yxg2ee) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2f90817b0569a18ced72bced36c78d45119baf4/pkg/storage/queue_test.go#L775)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This seems unrelated to this PR. Revert?
</blockquote></details>

It's related; see the commit message. The value of `v` was around 4.5ms on windows.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14597)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 3 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/util/timeutil/time_windows.go, line 28 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14597#-KgtdP-h3iNTA-niGCpE:-KgtdP-h3iNTA-niGCpF:b-dyyb6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b2f3e8ce9b7f601e62d921c7559fa51c9e4dbd0/pkg/util/timeutil/time_windows.go#L28)):*
> ```Go
> )
> 
> func now() time.Time {
> ```

Add a comment that we're doing this because it has higher resolution than `time.Now`, but it requires windows 8 (right?). We should also add a TODO and/or an issue to reexamine this in go 1.9 (since this will prevent us from getting the benefits of the planned monotonic clock support, and we may be able to switch back to `time.Now` at that point)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14597#-:-Kgtdkltopj0eYtKyv0V:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 4 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/util/timeutil/now_windows.go, line 28 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14597#-KgtdP-h3iNTA-niGCpE:-KgtjaTUWLYcYTTRDZd6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b2f3e8ce9b7f601e62d921c7559fa51c9e4dbd0/pkg/util/timeutil/now_windows.go#L28)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Add a comment that we're doing this because it has higher resolution than `time.Now`, but it requires windows 8 (right?). We should also add a TODO and/or an issue to reexamine this in go 1.9 (since this will prevent us from getting the benefits of the planned monotonic clock support, and we may be able to switch back to `time.Now` at that point)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14597)*
<!-- Sent from Reviewable.io -->

tamird committed:
Makefile: quote paths

In Windows, these paths include forward slashes, which the shell
eagerly interprets as escapes, but we need them to be literal, so
explicit quoting is required.

tamird committed:
Makefile: consistently include LINKFLAGS and allow in env

This is necessary to sanely work around https://github.com/golang/go/issues/19734.
